src/middleware.rs

```diff
@@ -134,20 +136,86 @@ where
         req.extensions_mut().insert(root_span_wrapper);
 
         let fut = root_span.in_scope(|| self.service.call(req));
-        Box::pin(
-            async move {
-                let outcome = fut.await;
+
+        TracingResponse {
+            fut,
+            span: root_span,
+            _root_span_type: std::marker::PhantomData,
+        }
+    }
+}
+
+#[doc(hidden)]
+#[pin_project::pin_project]
+pub struct TracingResponse<F, RootSpanType> {
+    #[pin]
+    fut: F,
+    span: Span,
+    _root_span_type: std::marker::PhantomData<RootSpanType>,
+}
+
+#[doc(hidden)]
+#[pin_project::pin_project]
+pub struct StreamSpan<B> {
+    #[pin]
+    body: B,
+    span: Span,
+}
+
+impl<F, B, RootSpanType> Future for TracingResponse<F, RootSpanType>
+where
+    F: Future<Output = Result<ServiceResponse<B>, Error>>,
+    B: actix_web::dev::MessageBody + 'static,
+    RootSpanType: RootSpanBuilder + Unpin,
```

Comment:
I had to assert RootSpanType is Unpin here, which I may have been able to get around by marking `_root_span_type` as `#[pin]`

That said, I can't imagine someone would make a `!Unpin` RootSpanBuilder, since a value of that type is never instantiated in the first place

Reply:
Sorry, I don't get it, could you please explain why you think the `Unpin` bound is necessary ? We don't ever pin it, actually we don't even use it, it's a `PhantomData`.

Reply:
Because it didn't compile without the Unpin bound :shrug: 

It seems like `PhantomData` is only `Unpin` when it's inner type is also `Unpin`
![image](https://user-images.githubusercontent.com/5304643/135148959-c2227f3d-21f6-4fc9-b6ef-0445138e678f.png)

Reply:
it could be safe to do something like
```
struct PhantomUnpin<T>(PhantomData<T>);
unsafe impl<T> Unpin for PhantomUnpin<T> {}
```
and then use the `PhantomUnpin` type instead

Reply:
Hmm, I can compile the crate just fine without the `Unpin` bound. Why do you need `_root_span_type` to be `Unpin` ? It shouldn't matter, since we never pin it, in fact, it won't even make a difference for either `TracingResponse` is `Unpin` or not, `pin_project` only takes `#[pin]` fields into account when implemeting `Unpin`.

Reply:
Hmm you're right. I don't know what I was doing wrong earlier that was causing those errors, then.